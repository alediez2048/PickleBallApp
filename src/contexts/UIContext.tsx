import React, { createContext, useContext, useReducer, useCallback, useEffect } from 'react';
import { ColorSchemeName, useColorScheme as useRNColorScheme } from 'react-native';

// State interface
interface UIState {
  colorScheme: ColorSchemeName;
  isLoading: boolean;
  toast: {
    visible: boolean;
    message: string;
    type: 'success' | 'error' | 'info';
  };
}

// Action types
type UIAction =
  | { type: 'SET_COLOR_SCHEME'; payload: ColorSchemeName }
  | { type: 'SET_LOADING'; payload: boolean }
  | { type: 'SHOW_TOAST'; payload: { message: string; type: 'success' | 'error' | 'info' } }
  | { type: 'HIDE_TOAST' };

// Initial state
const initialState: UIState = {
  colorScheme: 'light',
  isLoading: false,
  toast: {
    visible: false,
    message: '',
    type: 'info',
  },
};

// Reducer
function uiReducer(state: UIState, action: UIAction): UIState {
  switch (action.type) {
    case 'SET_COLOR_SCHEME':
      return { ...state, colorScheme: action.payload };
    case 'SET_LOADING':
      return { ...state, isLoading: action.payload };
    case 'SHOW_TOAST':
      return {
        ...state,
        toast: {
          visible: true,
          message: action.payload.message,
          type: action.payload.type,
        },
      };
    case 'HIDE_TOAST':
      return {
        ...state,
        toast: {
          ...state.toast,
          visible: false,
        },
      };
    default:
      return state;
  }
}

// Context interface
interface UIContextType extends UIState {
  setColorScheme: (scheme: ColorSchemeName) => void;
  setLoading: (isLoading: boolean) => void;
  showToast: (message: string, type: 'success' | 'error' | 'info') => void;
  hideToast: () => void;
  useThemeColor: (props: { light?: string; dark?: string }, colorName: string) => string;
}

// Create context
const UIContext = createContext<UIContextType | undefined>(undefined);

// Provider component
export function UIProvider({ children }: { children: React.ReactNode }) {
  const [state, dispatch] = useReducer(uiReducer, initialState);
  const systemColorScheme = useRNColorScheme();

  // Sync with system color scheme
  useEffect(() => {
    dispatch({ type: 'SET_COLOR_SCHEME', payload: systemColorScheme });
  }, [systemColorScheme]);

  const setColorScheme = useCallback((scheme: ColorSchemeName) => {
    dispatch({ type: 'SET_COLOR_SCHEME', payload: scheme });
  }, []);

  const setLoading = useCallback((isLoading: boolean) => {
    dispatch({ type: 'SET_LOADING', payload: isLoading });
  }, []);

  const showToast = useCallback((message: string, type: 'success' | 'error' | 'info') => {
    dispatch({ type: 'SHOW_TOAST', payload: { message, type } });
    // Auto-hide toast after 3 seconds
    setTimeout(() => {
      dispatch({ type: 'HIDE_TOAST' });
    }, 3000);
  }, []);

  const hideToast = useCallback(() => {
    dispatch({ type: 'HIDE_TOAST' });
  }, []);

  const useThemeColor = useCallback(
    (props: { light?: string; dark?: string }, colorName: string): string => {
      const theme = state.colorScheme ?? 'light';
      const colorFromProps = props[theme];

      if (colorFromProps) {
        return colorFromProps;
      }

      // Default theme colors
      const Colors = {
        light: {
          text: '#000000',
          background: '#ffffff',
          primary: '#2f95dc',
          secondary: '#f4f4f5',
          tint: '#2f95dc',
          tabIconDefault: '#cccccc',
          tabIconSelected: '#2f95dc',
        },
        dark: {
          text: '#ffffff',
          background: '#000000',
          primary: '#2f95dc',
          secondary: '#27272a',
          tint: '#ffffff',
          tabIconDefault: '#cccccc',
          tabIconSelected: '#ffffff',
        },
      };

      return Colors[theme][colorName as keyof typeof Colors.light];
    },
    [state.colorScheme]
  );

  const value = {
    ...state,
    setColorScheme,
    setLoading,
    showToast,
    hideToast,
    useThemeColor,
  };

  return <UIContext.Provider value={value}>{children}</UIContext.Provider>;
}

// Hook for using the context
export function useUI() {
  const context = useContext(UIContext);
  if (context === undefined) {
    throw new Error('useUI must be used within a UIProvider');
  }
  return context;
} 