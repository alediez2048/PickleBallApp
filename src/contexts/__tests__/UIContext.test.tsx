import React from 'react';
import renderer, { act } from 'react-test-renderer';
import { View, Text, Pressable } from 'react-native';
import { UIProvider, useUI } from '../UIContext';

// Mock timers for toast auto-hide behavior
jest.useFakeTimers();

// Test component that uses UI context
const TestComponent: React.FC = () => {
  const { colorScheme, isLoading, toast, setColorScheme, setLoading, showToast, hideToast, useThemeColor } = useUI();
  const backgroundColor = useThemeColor({ light: '#ffffff', dark: '#000000' }, 'background');

  return (
    <View style={{ backgroundColor }} testID="test-container">
      <Pressable testID="toggle-theme" onPress={() => setColorScheme(colorScheme === 'dark' ? 'light' : 'dark')}>
        <Text>Toggle Theme</Text>
      </Pressable>
      <Pressable testID="toggle-loading" onPress={() => setLoading(!isLoading)}>
        <Text>Toggle Loading</Text>
      </Pressable>
      <Pressable testID="show-toast" onPress={() => showToast('Test Toast', 'info')}>
        <Text>Show Toast</Text>
      </Pressable>
      <Pressable testID="hide-toast" onPress={hideToast}>
        <Text>Hide Toast</Text>
      </Pressable>
      <Text testID="color-scheme">{colorScheme}</Text>
      <Text testID="loading-state">{isLoading ? 'loading' : 'not-loading'}</Text>
      <Text testID="toast-state">{toast.visible ? 'visible' : 'hidden'}</Text>
      <Text testID="toast-message">{toast.message}</Text>
      <Text testID="toast-type">{toast.type}</Text>
      <Text testID="background-color">{backgroundColor}</Text>
    </View>
  );
};

describe('UIContext', () => {
  afterEach(() => {
    jest.clearAllTimers();
  });

  it('provides initial state', () => {
    const tree = renderer
      .create(
        <UIProvider>
          <TestComponent />
        </UIProvider>
      )
      .toJSON();
    
    expect(tree).toMatchSnapshot('initial state');
  });

  it('toggles loading state', () => {
    const component = renderer.create(
      <UIProvider>
        <TestComponent />
      </UIProvider>
    );

    // Get initial snapshot
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot('before loading');

    // Toggle loading state
    act(() => {
      const toggleButton = component.root.findByProps({ testID: 'toggle-loading' });
      toggleButton.props.onPress();
    });

    // Get updated snapshot
    tree = component.toJSON();
    expect(tree).toMatchSnapshot('after loading enabled');

    // Toggle loading state again
    act(() => {
      const toggleButton = component.root.findByProps({ testID: 'toggle-loading' });
      toggleButton.props.onPress();
    });

    // Get final snapshot
    tree = component.toJSON();
    expect(tree).toMatchSnapshot('after loading disabled');
  });

  it('shows and auto-hides toast', () => {
    const component = renderer.create(
      <UIProvider>
        <TestComponent />
      </UIProvider>
    );

    // Get initial snapshot
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot('before showing toast');

    // Show toast
    act(() => {
      const showToastButton = component.root.findByProps({ testID: 'show-toast' });
      showToastButton.props.onPress();
    });

    // Get snapshot after showing toast
    tree = component.toJSON();
    expect(tree).toMatchSnapshot('after showing toast');

    // Fast-forward toast auto-hide timeout
    act(() => {
      jest.advanceTimersByTime(3000);
    });

    // Get snapshot after toast auto-hidden
    tree = component.toJSON();
    expect(tree).toMatchSnapshot('after toast auto-hidden');
  });

  it('manually hides toast', () => {
    const component = renderer.create(
      <UIProvider>
        <TestComponent />
      </UIProvider>
    );

    // Show toast
    act(() => {
      const showToastButton = component.root.findByProps({ testID: 'show-toast' });
      showToastButton.props.onPress();
    });

    // Get snapshot after showing toast
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot('with visible toast');

    // Manually hide toast
    act(() => {
      const hideToastButton = component.root.findByProps({ testID: 'hide-toast' });
      hideToastButton.props.onPress();
    });

    // Get snapshot after manually hiding toast
    tree = component.toJSON();
    expect(tree).toMatchSnapshot('after manually hiding toast');
  });

  it('toggles color scheme', () => {
    const component = renderer.create(
      <UIProvider>
        <TestComponent />
      </UIProvider>
    );

    // Get initial snapshot
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot('with initial color scheme');

    // Toggle color scheme
    act(() => {
      const toggleThemeButton = component.root.findByProps({ testID: 'toggle-theme' });
      toggleThemeButton.props.onPress();
    });

    // Get snapshot after theme toggle
    tree = component.toJSON();
    expect(tree).toMatchSnapshot('after toggling color scheme');
  });
}); 