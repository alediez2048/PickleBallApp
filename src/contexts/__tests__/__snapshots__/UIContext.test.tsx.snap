// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UIContext provides initial state: initial state 1`] = `null`;

exports[`UIContext shows and auto-hides toast: before showing toast 1`] = `null`;

exports[`UIContext toggles color scheme: with initial color scheme 1`] = `null`;

exports[`UIContext toggles loading state: before loading 1`] = `null`;
